kong/tools/stream_api.lua

```diff
@@ -0,0 +1,77 @@
+
+require "lua_pack"
+
+
+local kong       = kong
+local st_pack    = string.pack      -- luacheck: ignore string
+local st_unpack  = string.unpack    -- luacheck: ignore string
+local st_format  = string.format
+
+local PREFIX = ngx.config.prefix()
+
+local stream_api = {}
+
+local _handlers  = {}
+
+
+function stream_api.register(k, f)
+  _handlers[k] = f
+end
+
+function stream_api.request(key, data, socket_path)
+  local socket = assert(ngx.socket.udp())
+  assert(socket:setpeername(socket_path or "unix:" .. PREFIX .. "/stream_rpc.sock"))
+
+  local ok, err = socket:send(st_pack("=PP", key, data))
+  if not ok then
+    socket:close()
+    return ok, err
+  end
+
+  data, err = socket:receive()
+  if not data then
+    socket:close()
+    return data, err
+  end
+
+  local _, status, payload = st_unpack(data, "=SP")
+  if status == 0 then
```

Comment:
It's not critical, but I do feel using `0` as the success code is a bit easier to remember. (I know OpenSSL is an exception, but neither Nginx nor typical C APIs the adopts `0` for error codes..) 😂 

Unless of course, that the `1` here is the number of return values like Lua APIs, then we can keep this format, but a comment would remind future readers of this.

Reply:
i was back and forth about one way or the other, but didn't find any important advantage on either side.
my reasoning was "0 -> false -> bad, 1 -> true -> good".

the executable return codes reasoning is like "there's a single good outcome, but many bad ones", but that means that I have to double- and triple-check any non-trivial logic in shell scripts.  that bothers me every day :woozy_face: 

on C APIs the reasoning is mostly the same "many different errors", but it's usually hidden in constants or enums, so it's not so weird that `xxx_OK == 0`  (still, i usually do `xxx_None = 0, xxx_Ok = 1, xxx_Err_yyy = 2, 3...`)

in this case, the specific error is carried on the payload, so i'm ok with having a single error condition, and leaned towards avoiding the shell awkwardness.  but in the end, in Lua 0 is truthy, so there's no danger of that.

changed.  even ended doing the smallest amount of "different errors", using 1=transport error (no handler) and 2=error from handler (they both end as `nil, <errmsg>`, so :man_shrugging: )